### PR TITLE
Align error handling behavior

### PR DESCRIPTION
**Pull request by oshadmi**
Maintain legacy behavior
And some tuning of new behavior

### CODE REVIEW

File: src/commands.rs [hunk 1]
```diff
@@ -398,15 +398,17 @@ impl<'a, V: SelectValue> KeyValue<'a, V> {
...
+        match self.get_first(path) {
+            Ok(first) => match first.get_type() {
+                SelectValueType::Object => Ok(ObjectLen::Len(first.len().unwrap())),
+                _ => Err(
+                    err_msg_json_expected("object", self.get_type(path).unwrap().as_str())
+                        .as_str()
+                        .into(),
+                ),
+            },
+            _ => Ok(ObjectLen::NoneExisting),
```
Comment (MeirShpilraien) [inline src/commands.rs:411]: Doesn't it make more sense that this is an error?

File: src/commands.rs [hunk 2]
```diff
@@ -1744,20 +1770,19 @@ fn json_obj_keys<M>(redis_key: &mut M::ReadHolder, path: &str) -> RedisResult
...
-                res.push(match v {
-                    Some(v) => v.keys().unwrap().collect::<Vec<&str>>().into(),
-                    _ => RedisValue::Null,
-                });
-            }
-            res.into()
+    let root = redis_key
+        .get_value()?
+        .ok_or_else(RedisError::nonexistent_key)?;
+    let res: RedisValue = {
```
Comment (MeirShpilraien) [inline src/commands.rs:1776]: Before we handled the case where root is None, don't we need to handle it anymore?

File: src/commands.rs (hunk 1 above)
Comment (oshadmi) [reply src/commands.rs:411]: Yes, in new behavior it is an error (see `json_obj_len`). This is to maintain legacy behavior.

File: tests/pytest/test.py [hunk 3]
```diff
@@ -745,7 +745,7 @@ def testLenCommands(env):
     r.assertEqual(r.execute_command('JSON.ARRLEN', 'test', '.arr'), 6)
 
     # test elements with undefined lengths
-    r.assertEqual(r.execute_command('JSON.ARRLEN', 'test', '.bool'), None)
+    r.expect('JSON.ARRLEN', 'test', '.bool').raiseError()
```
Comment (MeirShpilraien) [inline tests/pytest/test.py:748]: Let add `.contains` with a substring we expect to see in the error

File: src/commands.rs (hunk 2 above)
Comment (oshadmi) [reply src/commands.rs:1776]: Now it is handled in `json_obj_keys` and `json_obj_keys_legacy`

File: tests/pytest/test_multi.py [hunk 4]
```diff
@@ -680,7 +679,8 @@ def testObjLenCommand(env):
     r.assertEqual(res, None)
 
     # Test missing path
-    r.expect('JSON.OBJLEN', 'doc1', '.nowhere').raiseError()
+    res = r.execute_command('JSON.OBJLEN', 'doc1', '.nowhere')
```
Comment (MeirShpilraien) [inline tests/pytest/test_multi.py:682]:
```suggestion
    r.expect('JSON.OBJLEN', 'doc1', '.nowhere').equal(None)
```

File: tests/pytest/test_multi.py [hunk 5]
```diff
@@ -680,7 +679,8 @@ def testObjLenCommand(env):
...
+    r.assertEqual(res, None)
```
Comment (MeirShpilraien) [inline tests/pytest/test_multi.py:683]:
```suggestion
```

File: tests/pytest/test_multi.py [hunk 6]
```diff
@@ -978,139 +978,316 @@ def testErrorMessage(env):
     res = r.execute_command('JSON.GET', 'doc1', '$')
     r.assertEqual([types_data], json.loads(res))
 
+    r.assertEqual(r.execute_command('HSET', 'hash_key', 'a', '1', 'b', '2'), 2)
```
Comment (MeirShpilraien) [inline tests/pytest/test_multi.py:981]:
```suggestion
    r.expect('HSET', 'hash_key', 'a', '1', 'b', '2').equal(2)
```

File: tests/pytest/test_multi.py [hunk 7]
```diff
@@ -978,139 +978,316 @@ def testErrorMessage(env):
...
     r.assertEqual(r.execute_command('JSON.OBJKEYS', 'doc1', '$.string'), [None])
     r.expect('JSON.OBJKEYS', 'doc1', '$.nowhere').raiseError().contains("does not exist")
-    r.assertEqual(r.execute_command('JSON.OBJKEYS', 'doc_none', '$.string'), None)
+    r.expect('JSON.OBJKEYS', 'doc_none', '$.string').raiseError().contains("doesn't exist")
+    r.expect('JSON.OBJKEYS', 'hash_key', '$.string').raiseError().contains("wrong Redis type")
 
-    r.expect('JSON.OBJKEYS', 'doc1', '.string').raiseError().contains("expected object but found string")
-    r.expect('JSON.OBJKEYS', 'doc1', '.nowhere').raiseError().contains("does not exist")
+    r.expect('JSON.OBJKEYS', 'doc1', '.string').raiseError().contains("not an object")
+    r.assertEqual(r.execute_command('JSON.OBJKEYS', 'doc1', '.nowhere'), None)
```
Comment (MeirShpilraien) [inline tests/pytest/test_multi.py:1112]:
```suggestion
    r.expect('JSON.OBJKEYS', 'doc1', '.nowhere').equal(None)
```

File: tests/pytest/test_multi.py [hunk 8]
```diff
@@ -978,139 +978,316 @@ def testErrorMessage(env):
...
     # OBJLEN
     r.assertEqual(r.execute_command('JSON.OBJLEN', 'doc1', '$.string'), [None])
     r.expect('JSON.OBJLEN', 'doc1', '$.nowhere').raiseError().contains("does not exist")
-    r.assertEqual(r.execute_command('JSON.OBJLEN', 'doc_none', '$.string'), None)
+    r.expect('JSON.OBJLEN', 'doc_none', '$.string').raiseError().contains("does not exist")
+    r.expect('JSON.OBJLEN', 'hash_key', '.string').raiseError().contains("wrong Redis type")
 
     r.expect('JSON.OBJLEN', 'doc1', '.boolean').raiseError().contains("expected object but found boolean")
-    r.expect('JSON.OBJLEN', 'doc1', '.nowhere').raiseError().contains("does not exist")
+    r.assertEqual(r.execute_command('JSON.OBJLEN', 'doc1', '.nowhere'), None)
```
Comment (MeirShpilraien) [inline tests/pytest/test_multi.py:1132]:
```suggestion
    r.expect('JSON.OBJLEN', 'doc1', '.nowhere').equal(None)
```

**Review by MeirShpilraien — COMMENTED** [review]
Looks good, some comments.
Also why did you updated readies submodule?

File: tests/pytest/test_multi.py (hunk 4 above)
Comment (oshadmi) [reply tests/pytest/test_multi.py:682]: Keeping `execute_command` to handle cluster connection.

File: tests/pytest/test_multi.py (hunk 6 above)
Comment (oshadmi) [reply tests/pytest/test_multi.py:981]: Keeping `execute_command` to handle cluster connection

File: tests/pytest/test_multi.py (hunk 7 above)
Comment (oshadmi) [reply tests/pytest/test_multi.py:1112]: Keeping `execute_command` to handle cluster connection

File: tests/pytest/test_multi.py (hunk 8 above)
Comment (oshadmi) [reply tests/pytest/test_multi.py:1132]: Keeping `execute_command` to handle cluster connection

**Comment by oshadmi** [comment]
> Looks good, some comments. Also why did you updated readies submodule?

Build was [failing](https://app.circleci.com/pipelines/github/RedisJSON/RedisJSON/1880/workflows/6787f4e6-8d44-4238-b644-a2df134942c9/jobs/4558?invite=true#step-104-128)

File: src/commands.rs (hunk 1 above)
Comment (MeirShpilraien) [reply src/commands.rs:411]: So why add it as part of the valid return value and not as error?

**Review by MeirShpilraien — APPROVED** [review]
Added one followup question on one of the comment, but no a blocker.

File: src/commands.rs (hunk 1 above)
Comment (oshadmi) [reply src/commands.rs:411]: Just for backward compatibility with V 1.0 (only with legacy path)
`JSON.OBJKEYS doc1 .nowhere` and `JSON.OBJKEYS no_such_doc .` should return null (as before)

**Review by oshadmi — COMMENTED** [review]
Thanks!

**Pull request merged**